### PR TITLE
 Mark v1alpha1 source as deprecated in conversion

### LINKED ISSUE #1547: Mark v1alpha1 being Deprecated
**Problem**
We decided to remove v1alpha1 support after 0.17 cuts. We should document in the release notes that we will be removing it. Further, we should have a deprecation warning in the status of the resources, probably set in the v1alpha1 conversion logic.

**[Persona:](https://github.com/knative/eventing/blob/master/docs/personas.md)** 
Developer

**Exit Criteria**
Mark v1alpha1 to be deprecated and state in the release notes that this will be remove in 0.18.

**Additional context (optional)**
See https://github.com/google/knative-gcp/issues/1544
and https://github.com/google/knative-gcp/issues/1435

### PR DESCRIPTION
**Pull request by danyinggu**
Fixes #1547 

## Proposed Changes

- Mark v1alpha1 as deprecated Condition in Status when doing the conversion.
- Refactor channel conversion UTs and nits.

Currently, the behaviours are: 
If describe v1alpha1 source, there will be a deprecated warning condition.
``` 
kubectl describe cloudauditlogssources.v1alpha1.events.cloud.google.com
Name:         cloudauditlogssource-test
Namespace:    default
Labels:       <none>
Annotations:  cluster-name: cluster-4
API Version:  events.cloud.google.com/v1alpha1
Kind:         CloudAuditLogsSource
Metadata:
  Creation Timestamp:  2020-08-07T22:44:24Z
  Finalizers:
    cloudauditlogssources.events.cloud.google.com
  Generation:        1
  Resource Version:  41882670
  Self Link:         /apis/events.cloud.google.com/v1alpha1/namespaces/default/cloudauditlogssources/cloudauditlogssource-test
  UID:               d3209a03-a4d0-476a-8642-7d116018b6b8
Spec: ...
Status:
  Conditions:
    Last Transition Time:  2020-08-07T22:44:54Z
    Message:               V1alpha1 Sources will be deprecated after 0.18 cut.
    Reason:                v1alpha1Deprecated
    Severity:              Warning
    Status:                True
    Type:                  Deprecated
    Last Transition Time:  2020-08-07T22:44:31Z
    Status:                True
    Type:                  PullSubscriptionReady
    Last Transition Time:  2020-08-07T22:44:31Z
    Status:                True
    Type:                  Ready
    Last Transition Time:  2020-08-07T22:44:31Z
    Status:                True
    Type:                  SinkReady
    Last Transition Time:  2020-08-07T22:44:27Z
    Status:                True
    Type:                  TopicReady
  Observed Generation:     1
  Project Id:              danyinggu-knative-dev
  Sink Uri:                http://event-display.default.svc.cluster.local/
  Stackdriver Sink:        cre-src_default_cloudauditlogssource-test_d3209a03-a4d0-476a-8642-7d116018b6b8
  Subscription Id:         cre-src_default_cloudauditlogssource-test_83e47966-abf4-4a51-8f2a-c709a02a1d22
  Topic Id:                cre-src_default_cloudauditlogssource-test_d3209a03-a4d0-476a-8642-7d116018b6b8

```

While, when
```
kubectl describe cloudauditlogssources.v1.events.cloud.google.com
Name:         cloudauditlogssource-test
Namespace:    default
Labels:       <none>
Annotations:  API Version:  events.cloud.google.com/v1
Kind:         CloudAuditLogsSource
Metadata:
  Creation Timestamp:  2020-08-07T22:44:24Z
  Finalizers:
    cloudauditlogssources.events.cloud.google.com
  Generation:        1
  Resource Version:  41882670
  Self Link:         /apis/events.cloud.google.com/v1/namespaces/default/cloudauditlogssources/cloudauditlogssource-test
  UID:               d3209a03-a4d0-476a-8642-7d116018b6b8
Spec:
  Method Name:  google.pubsub.v1.Publisher.CreateTopic
  Secret:
    Key:         key.json
    Name:        google-cloud-key
  Service Name:  pubsub.googleapis.com
  Sink:
    Ref:
      API Version:  v1
      Kind:         Service
      Name:         event-display
Status:
  Conditions:
    Last Transition Time:  2020-08-07T22:44:31Z
    Status:                True
    Type:                  PullSubscriptionReady
    Last Transition Time:  2020-08-07T22:44:31Z
    Status:                True
    Type:                  Ready
 Last Transition Time:  2020-08-07T22:44:31Z
    Status:                True
    Type:                  SinkReady
    Last Transition Time:  2020-08-07T22:44:27Z
    Status:                True
    Type:                  TopicReady
  Observed Generation:     1
```

There is no warning Deprecated condition.

**Release Note**

```release-note
v1alpha1 resources (Source, Topic, PullSubscription and Channel) have been deprecated and will be removed in 0.18.
Please use v1 (v1beta1 for Channel) instead.
```

**Docs**

### CODE REVIEW

**Comment by danyinggu** [comment]
@nachocano @capri-xiyue This is a WIP prototype. I only made the change in CloudAuditLogsSource for now. Could you please take a look to see if that is what we expect? If so, I will propagate it to the other sources.
(The conversion_helper UTs are missing, if the logic is okay, then I will add them.)

File: pkg/apis/convert/conversion_helper.go [hunk 1]
```diff
@@ -19,17 +19,33 @@ package convert
...
 	eventingduckv1beta1 "knative.dev/eventing/pkg/apis/duck/v1beta1"
+	"knative.dev/pkg/apis"
+	pkgduckv1 "knative.dev/pkg/apis/duck/v1"
 	pkgduckv1alpha1 "knative.dev/pkg/apis/duck/v1alpha1"
 	pkgduckv1beta1 "knative.dev/pkg/apis/duck/v1beta1"
 )
 
+const (
+	deprecatedV1Alpha1Reason = "v1alpha1Deprecated"
+	deprecatedV1Alpha1Msg    = "V1alpha1 Sources will be deprecated after 0.18 cut."
```
Comment (nachocano) [inline pkg/apis/convert/conversion_helper.go:36]: maybe something like: v1alpha has been deprecated and will be removed in 0.18. WDYT?

File: pkg/apis/convert/conversion_helper.go [hunk 2]
```diff
@@ -309,3 +325,15 @@ func FromV1beta1SubscribableSpec(from *eventingduckv1beta1.SubscribableSpec) *ev
 	}
 	return &to
 }
+
+// A helper function to mark v1alpha1 Deprecated Condition in the Status.
+// We mark the Condition in status during conversion from a higher version to v1alpha1.
+func MarkV1alpha1Deprecated(cs *apis.ConditionSet, s *pkgduckv1.Status) {
```
Comment (nachocano) [inline pkg/apis/convert/conversion_helper.go:331]: I'd suggest you also add another comment to this and the remove method saying something like:
`//TODO remove after the 0.17 cut.`

**Comment by nachocano** [comment]
> @nachocano @capri-xiyue This is a WIP copy. I only made the change in CloudAuditLogsSource for now. Could you please take a look to see if this logic is what we expect or if there is anything missing? If this is what we want, I will propagate it to the other sources.
> (The conversion_helper UTs are missing, just want to check if the logic is okay before adding them.)

I think this is what we want. Just added two cosmetic comments. Thanks @danyinggu !!!
Note that we should do a similar thing for the channel as well....

**Comment by danyinggu** [comment]
> > @nachocano @capri-xiyue This is a WIP copy. I only made the change in CloudAuditLogsSource for now. Could you please take a look to see if this logic is what we expect or if there is anything missing? If this is what we want, I will propagate it to the other sources.
> > (The conversion_helper UTs are missing, just want to check if the logic is okay before adding them.)
> 
> I think this is what we want. Just added two cosmetic comments. Thanks @danyinggu !!!
> Note that we should do a similar thing for the channel as well....

Thank you so much for taking a look at this! 
Sounds good, I will promote this logic (with the suggested fixes) to all other sources and channel as well :)

File: pkg/apis/convert/conversion_helper.go (hunk 1 above)
Comment (danyinggu) [reply pkg/apis/convert/conversion_helper.go:36]: Yes, this is more explicit. I will change it.

**Comment by nachocano** [comment]
> > > @nachocano @capri-xiyue This is a WIP copy. I only made the change in CloudAuditLogsSource for now. Could you please take a look to see if this logic is what we expect or if there is anything missing? If this is what we want, I will propagate it to the other sources.
> > > (The conversion_helper UTs are missing, just want to check if the logic is okay before adding them.)
> > 
> > 
> > I think this is what we want. Just added two cosmetic comments. Thanks @danyinggu !!!
> > Note that we should do a similar thing for the channel as well....
> 
> Thank you so much for taking a look at this!
> Sounds good, I will promote this logic (with the suggested fixes) to all other sources and channel as well :)

Great! Thanks @danyinggu ! And one last thing, if you can add a RELEASE NOTES to the description of the PR it'd be great. Saying something that v1alpha1 resources have been deprecated and will be removed in 0.18. Please use v1 instead.

**Comment by nachocano** [comment]
/lgtm
/approve

Thanks @danyinggu !!!

**Pull request merged**